diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9e3b555..b0abd84 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3069,7 +3069,7 @@ if(RPMBUILD_EXECUTABLE AND GIT_EXECUTABLE)
 	endif()
 
 	execute_process(
-		COMMAND git describe --abbrev=8 --match v[1-9]*
+		COMMAND git describe --always --abbrev=8 --match v[1-9]*
 		OUTPUT_VARIABLE _git_description
 		OUTPUT_STRIP_TRAILING_WHITESPACE
 		WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
diff --git a/wsutil/CMakeLists.txt b/wsutil/CMakeLists.txt
index 0367cd1..6382a2c 100644
--- a/wsutil/CMakeLists.txt
+++ b/wsutil/CMakeLists.txt
@@ -69,6 +69,7 @@ set(WSUTIL_PUBLIC_HEADERS
 	ws_mempbrk_int.h
 	ws_pipe.h
 	ws_printf.h
+	wstmpdir.h
 	wsjsmn.h
 	xtea.h
 )
@@ -118,6 +118,7 @@ set(WSUTIL_COMMON_FILES
 	unicode-utils.c
 	ws_mempbrk.c
 	ws_pipe.c
+	wstmpdir.c
 	wsgcrypt.c
 	wsjsmn.c
 	xtea.c
